### PR TITLE
feature(buttons): add button of new UI

### PR DESCRIPTION
**Pull request by julianosbento**
# What

Add button of new UI

# Why

Create new possibilities of interaction based on new UI

# How

- Remove legacy code
- Add new buttons and utils for buttons

# Sample

# QA

Run application locally and test the props like `variant`, `type`, `size`,  `fill`, `loading`, `disabled` and others

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,9 +1,20 @@
 module.exports = {
-  setupFilesAfterEnv: ['<rootDir>setupJest.ts', '@testing-library/jest-native/extend-expect'],
```
Comment (murillo94) [inline jest.config.js:2]: what's the main reason to remove it? the lib doesn't recommend it anymore?

File: package.json [hunk 2]
```diff
@@ -89,16 +90,11 @@
   },
   "dependencies": {
     "@magnetis/astro-galaxy-tokens": "^1.2.3",
+    "@magnetis/astro-tokens": "^0.0.8",
     "@react-native-community/slider": "^3.0.3",
     "react-native-svg": "^12.1.0",
     "y18n": "5.0.8"
   },
-  "husky": {
```
Comment (murillo94) [inline package.json:96]: no more husky 😮 

it's worth remove husky lib too?

**Review by murillo94 — APPROVED** [review]

**Review by an-mag — APPROVED** [review]

File: jest.config.js (hunk 1 above)
Comment (julianosbento) [reply jest.config.js:2]: We didn't remove it, we changed for `setupFilesAfterEnv: ['<rootDir>setupJest.ts'],`

File: package.json (hunk 2 above)
Comment (julianosbento) [reply package.json]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/628b353398a34357662d0c48fb08103c4706dfee

Comment (lucianomlima) [reply package.json]: Guys, husky configuration is now in directory `.husky`. The config on package.json is deprecated.

File: src/tokens/fonts.ts [hunk 3]
```diff
@@ -1,8 +1,18 @@
-export const poppins = 'Poppins-Regular';
-export const poppinsMedium = 'Poppins-Medium';
-export const poppinsSemiBold = 'Poppins-SemiBold';
-export const poppinsBold = 'Poppins-Bold';
-export const poppinsBlack = 'Poppins-Black';
+import { typography } from '@magnetis/astro-tokens';
+import { ButtonSize } from '../components/Buttons/types';
```
Comment (lucianomlima) [inline src/tokens/fonts.ts:2]: O que acha de ao invés de usar o ButtonSize, criar um type para a fonte, por que evita esse acoplamento e ainda já deixa os types relacionados a tamanho de fonte declarados.

File: src/tokens/fonts.ts [hunk 4]
```diff
@@ -1,8 +1,18 @@
...
-export const lato = 'Lato-Regular';
-export const latoBold = 'Lato-Bold';
+export type FontSize = {
+  small: typeof typography.fontSizeMicro;
+  medium: typeof typography.fontSizeMini;
+  large: typeof typography.fontSizeSmaller;
+  xlarge: typeof typography.fontSizeMedium;
+};
+
+export function getFontSize(size: ButtonSize): number {
```
Comment (lucianomlima) [inline src/tokens/fonts.ts:11]: Aqui ao invés de retornar number, seria melhor retornar os valores que são possíveis para fonte.
Dessa forma deixa mais restrito e avisará quando um valor for usado e não fizer parte dos tokens de tamanho de tipografia.

File: storybook/options.ts [hunk 5]
```diff
@@ -1,12 +1,11 @@
 import { Colors, colors } from '@magnetis/astro-galaxy-tokens';
 import gradients from '@tokens/gradients';
-import { Size } from '@tokens/sizes';
 import * as icons from '@components/Icons';
 import type { IconID } from '@components/Icons';
-import { ButtonColor } from '@components/Buttons';
+import { ButtonSize, ButtonType, ButtonVariant } from '@components/Buttons/types';
```
Comment (lucianomlima) [inline storybook/options.ts:5]:
```suggestion
import type { ButtonSize, ButtonType, ButtonVariant } from '@components/Buttons/types';
```

File: storybook/options.ts [hunk 6]
```diff
@@ -15,7 +14,7 @@ export const colorOptionsWithTransparent: { [key: string]: Colors[keyof Colors]
   transparent: 'transparent',
 };
 
-export const sizeOptions: Size[] = ['very-small', 'small', 'medium', 'large'];
+export const sizeOptions: ButtonSize[] = ['small', 'medium', 'large', 'xlarge'];
```
Comment (lucianomlima) [inline storybook/options.ts:17]: Essas options poderiam estar declaradas próximo ao componente e serem reutilizadas em todos os lugares que precisa ter os valores permitidos não?
Inclusive, daria para usar esses valores para gerar o type que garantisse que não poderia ser passado outra informação.

Dessa forma, ao adicionar um novo valor, ou remover um novo valor, o type já teria o reflexo da mudança.

File: storybook/options.ts [hunk 7]
```diff
@@ -27,6 +26,8 @@ export const gradientOptions = Object.keys(gradients).reduce(
 
 export const iconOptions: Array<IconID> = Object.keys(icons).map((key) => key.replace('Icon', '') as IconID);
 
-export const buttonColorOptions: ButtonColor[] = ['uranus', 'venus', 'mars', 'earth'];
+export const buttonVariantOptions: ButtonVariant[] = ['primary', 'secondary', 'alert', 'success', 'inverted'];
+
+export const buttonTypeOptions: ButtonType[] = ['solid', 'subtle', 'outline', 'ghost'];
```
Comment (lucianomlima) [inline storybook/options.ts:31]: O mesmo ponto sobre os valores de tamanho.

**Review by lucianomlima — COMMENTED** [review]
Não olhei tudo, mas encontrei vários pontos de melhoria. Veja o que acha 😉

File: src/components/Buttons/BaseButton.tsx [hunk 8]
```diff
@@ -67,11 +65,11 @@ function BaseButton({
   }
 
   const pressableProps = {
-    onPress: props.onPress,
-    disabled: disabled,
+    disabled,
+    hitSlop: typeof props.hitSlop !== 'undefined' ? props.hitSlop : hitSlop,
```
[the remark this answers is not in the document]
Comment (julianosbento) [reply src/components/Buttons/BaseButton.tsx:69]: How, this is a definition, that we are already using at the buttons, we could hold this definitions, and after team definitions, we change the hitSlop, could be?

[the remark this answers is not in the document]
Comment (lucianomlima) [reply src/components/Buttons/BaseButton.tsx:69]: Yeah, sure!

File: storybook/options.ts (hunk 7 above)
Comment (julianosbento) [reply storybook/options.ts:31]: Answer [here](https://github.com/magnetis/astro-native/pull/265#discussion_r707429972)

File: src/components/Buttons/BaseButton.tsx [hunk 9]
```diff
@@ -1,37 +1,38 @@
...
 
 interface BaseButtonProps extends PressableProps {
-  children: ReactElement;
-  testID?: string;
   accessibilityLabel?: string;
-  loading?: boolean;
-  disabled?: boolean;
   activityIndicatorColor: string;
-  onPress: () => void;
+  children: ReactElement;
```
[the remark this answers is not in the document]
Comment (julianosbento) [reply src/components/Buttons/BaseButton.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: src/components/Buttons/BaseButton.tsx [hunk 10]
```diff
@@ -67,11 +65,11 @@ function BaseButton({
...
 
   const pressableProps = {
-    onPress: props.onPress,
-    disabled: disabled,
+    disabled,
+    hitSlop: typeof props.hitSlop !== 'undefined' ? props.hitSlop : hitSlop,
     style: [styles.button, computedStyles],
     testID: props.testID,
-    hitSlop: typeof props.hitSlop !== 'undefined' ? props.hitSlop : hitSlop,
+    onPress: props.onPress,
```
[the remark this answers is not in the document]
Comment (julianosbento) [reply src/components/Buttons/BaseButton.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: src/components/Buttons/types.ts [hunk 11]
```diff
@@ -1,40 +1,45 @@
...
 
 export interface ButtonProps extends PressableProps {
-  /** Text to be shown inside the button */
-  text: string;
-  /** Color to be used as background color of button. Defaults to `"uranus"`. */
-  color?: ButtonColor;
-  /** Button press callback */
-  onPress: () => void;
+  /** Suppress any user interaction with component. Defaults to `false`. */
+  disabled?: boolean;
```
[the remark this answers is not in the document]
Comment (julianosbento) [reply src/components/Buttons/types.ts]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: src/components/Buttons/types.ts [hunk 12]
```diff
@@ -1,40 +1,45 @@
...
-  /** Button fills entire parent when set to true. Defaults to `false`. */
-  fill?: boolean;
+  /** Text to be shown inside the button */
+  text: string;
+  /** Type of button, the default value is `"solid"` */
+  type: ButtonType;
+  /** Variant of button, the default value is `"primary"` */
+  variant: ButtonVariant;
+  /** Button press callback */
+  onPress: () => void;
```
[the remark this answers is not in the document]
Comment (julianosbento) [reply src/components/Buttons/types.ts]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: src/tokens/fonts.ts (hunk 3 above)
Comment (julianosbento) [reply src/tokens/fonts.ts]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: storybook/options.ts (hunk 5 above)
Comment (julianosbento) [reply storybook/options.ts]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/90c613e1a4c1f7b1df1030c6ef9820b02be906fd

File: storybook/options.ts (hunk 6 above)
Comment (lucianomlima) [reply storybook/options.ts]: I think I'm not clear but I try again. My suggestion it's not related with enum, it's related to get the types from the values. See the code above:
```ts
export type NavigationThemes = keyof typeof navigationThemes;

export const navigationThemes = {
  dark: {
    headerTintColor: colors.space100,
    headerBackgroundColor: colors.moon1000,
  },
  light: {
    headerTintColor: colors.moon1000,
    headerBackgroundColor: colors.space100,
  },
};
```
The `NavigationThemes` it's a pure type but is inferred by typescript from the values.
This means that if we change the const `navigationThemes` to another structure like:
```ts
export const navigationThemes = {
  black: {
    headerTintColor: colors.space100,
    headerBackgroundColor: colors.moon1000,
  },
  white: {
    headerTintColor: colors.moon1000,
    headerBackgroundColor: colors.space100,
  },
};
```
The type will be changed automatically. This way we have better and smart types that reduces our effort when changing code.
Moving the types to nearby the components, allow to infer them from code and import from the source of truth place.

File: src/components/Buttons/Button.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Primary buttons have a colored background.
+ *
+ *  Use these for the main action in a section or screen. You should display only one primary button at a time.
+ *
+ */
+function Button({
+  accessibilityLabel = '',
+  disabled = false,
+  fill = false,
```
Comment (lucianomlima) [inline src/components/Buttons/Button.tsx:20]: Optional props already have false value.
```suggestion
  disabled,
  fill,
```

File: src/components/Buttons/Button.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    backgroundColor,
+    borderColor: type === 'outline' ? textColor : backgroundColor,
+    disabled,
+    fill,
+    loading,
+    testID,
+    textColor,
+    size,
+    onPress,
+  };
```
Comment (lucianomlima) [inline src/components/Buttons/Button.tsx:48]: These props (accessibilityLabel, fill, loading, testID, onPress) already come from `...props` and are not used here. So you can omit them here and in destructuring.
```suggestion
  const baseProps = {
    activityIndicatorColor: textColor,
    backgroundColor,
    borderColor: type === 'outline' ? textColor : backgroundColor,
    disabled,
    textColor,
    size,
  };
```

File: src/components/Buttons/IconButton.tsx [hunk 15]
```diff
@@ -1,51 +1,59 @@
...
 /**
- * Icon buttons can be used for simple microinteractions and commands.
+ * Primary buttons have a colored background.
+ *
+ *  Use these for the main action in a section or screen. You should display only one primary button at a time.
+ *
  */
-const IconButton: React.FC<IconButtonProps> = ({
-  onPress,
+function IconButton({
```
Comment (lucianomlima) [inline src/components/Buttons/IconButton.tsx:16]: The same points that I put for Button

File: src/components/Buttons/__tests__/BaseButtonText.test.tsx [hunk 16]
```diff
@@ -1,25 +1,25 @@
...
+import { colors, typography } from '@magnetis/astro-tokens';
 
 import BaseButtonText from '../BaseButtonText';
 
 describe('BaseButtonText', () => {
   it('renders correctly with default props', () => {
-    const { getByText } = render(<BaseButtonText color={colors.uranus500}>text</BaseButtonText>);
+    const { getByText } = render(<BaseButtonText color={colors.solidPrimaryMedium}>text</BaseButtonText>);
 
     expect(getByText('text').props.style[0]).toEqual(
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/BaseButtonText.test.tsx:11]: You can use instead
```tsx
expect(getByText('text')).toHaveStyle({})
```

File: src/components/Buttons/__tests__/Button.test.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+      })
+    );
+
+    expect(button.props.style[1]).toEqual(
+      expect.objectContaining({
+        fontSize: typography.fontSizeMini,
+        lineHeight: typography.fontSizeMini * 1.5,
+        color: colors.solidBrightLightest,
+      })
+    );
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/Button.test.tsx:38]: I think can be done in one assert with `toHaveStyle`
```suggestion
    expect(button).toHaveStyle({
      fontFamily: typography.fontFamilyBase,
      alignItems: 'center',
      fontSize: typography.fontSizeMini,
      lineHeight: typography.fontSizeMini * 1.5,
      color: colors.solidBrightLightest,
    });
```

File: src/components/Buttons/__tests__/IconButton.test.tsx [hunk 18]
```diff
@@ -1,155 +1,35 @@
...
+import { colors } from '@magnetis/astro-tokens';
 
 import IconButton from '../IconButton';
-
-const onPress = jest.fn();
-
-const props = {
-  onPress,
-  testID: 'IconButton',
+import { IconButtonProps } from '..';
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/IconButton.test.tsx:6]:
```suggestion
import type { IconButtonProps } from '..';
```

File: src/components/Buttons/utils.ts [hunk 19]
```diff
@@ -1,136 +1,52 @@
...
+
   const regularVerticalPadding = {
-    'very-small': 4,
-    small: 8,
-    medium: 8,
-    large: 8,
+    small: 10,
+    medium: 17,
+    large: 20,
+    xlarge: 22,
```
Comment (lucianomlima) [inline src/components/Buttons/utils.ts:33]: Por que não usou os tokens aqui?

File: src/components/types.ts [hunk 20]
```diff
@@ -1,3 +1,5 @@
 import type { Insets } from 'react-native';
 
 export type HitSlop = null | Insets | number;
+
+export type Size = 'small' | 'medium' | 'large' | 'xlarge';
```
Comment (lucianomlima) [inline src/components/types.ts:5]: Here shouldn't be `ButtonSize`?
Maybe can be moved to `src/components/Buttons/types` to.

File: src/components/Buttons/utils.ts (hunk 19 above)
Comment (julianosbento) [reply src/components/Buttons/utils.ts:33]: Porque é um cálculo para definir o valor do padding

File: src/components/types.ts (hunk 20 above)
Comment (julianosbento) [reply src/components/types.ts:5]: I am using this for fonts and buttons, so I decided to abstract the level of the type

File: src/components/Buttons/__tests__/Button.test.tsx (hunk 17 above)
Comment (julianosbento) [reply src/components/Buttons/__tests__/Button.test.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/__tests__/IconButton.test.tsx (hunk 18 above)
Comment (julianosbento) [reply src/components/Buttons/__tests__/IconButton.test.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/__tests__/BaseButtonText.test.tsx (hunk 16 above)
Comment (julianosbento) [reply src/components/Buttons/__tests__/BaseButtonText.test.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/IconButton.tsx (hunk 15 above)
Comment (julianosbento) [reply src/components/Buttons/IconButton.tsx:16]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/Button.tsx (hunk 14 above)
Comment (julianosbento) [reply src/components/Buttons/Button.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/Button.tsx (hunk 13 above)
Comment (julianosbento) [reply src/components/Buttons/Button.tsx]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: storybook/options.ts (hunk 6 above)
Comment (julianosbento) [reply storybook/options.ts]: Fixed https://github.com/magnetis/astro-native/pull/265/commits/01ad556544c0cc2de509d82ac90cf681a2306b88

File: src/components/Buttons/utils.ts (hunk 19 above)
Comment (lucianomlima) [reply src/components/Buttons/utils.ts:33]: I think that what need to be done here it's not having a vertical padding defined. Like all the buttons have fixed height, you only need to aligns children on center. 
This way the padding will be adaptive to the content. What happens if we have a button with label 2 lines long?

File: src/components/types.ts (hunk 20 above)
Comment (lucianomlima) [reply src/components/types.ts]: But we will have a common `Size` type used for widths, heights and spacing.

File: src/components/Buttons/__tests__/BaseButton.test.tsx [hunk 21]
```diff
@@ -26,76 +26,66 @@ describe('BaseButton', () => {
...
+    expect(getByTestId('BaseButton')).toHaveStyle({
+      borderRadius: radius.small,
+      borderWidth: borders.hairline,
+      paddingVertical: 17,
+      paddingHorizontal: sizes.tiny,
+      backgroundColor: colors.solidPrimaryMedium,
+      borderColor: colors.solidPrimaryMedium,
+    });
+
+    expect(getByTestId('BaseButton.ActivityIndicator').props.color).toEqual(colors.solidBrightLightest);
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/BaseButton.test.tsx:73]: Can be use `toHaveProp` here
```ts
expect(getByTestId('BaseButton.ActivityIndicator')).toHaveProp('color', colors.solidBrightLightest);

File: src/components/Buttons/__tests__/Button.test.tsx [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    expect(button).toHaveStyle({
+      fontFamily: typography.fontFamilyBase,
+      alignItems: 'center',
+    });
+
+    expect(button).toHaveStyle({
+      fontSize: typography.fontSizeMini,
+      lineHeight: typography.fontSizeMini * 1.5,
+      color: colors.solidBrightLightest,
+    });
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/Button.test.tsx:34]: Can be a single assert

File: src/components/Buttons/__tests__/BaseButtonText.test.tsx [hunk 23]
```diff
@@ -1,26 +1,22 @@
...
-        fontSize: 16,
-        lineHeight: 16 * 1.5,
-        color: colors.uranus500,
-      })
-    );
+    expect(getByText('text')).toHaveStyle({
+      fontSize: typography.fontSizeMini,
+      lineHeight: typography.fontSizeMini * 1.5,
+      color: colors.solidPrimaryMedium,
+    });
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/BaseButtonText.test.tsx:20]: Can be a single assert

File: src/components/Buttons/__tests__/IconButton.test.tsx [hunk 24]
```diff
@@ -1,155 +1,35 @@
...
-        borderWidth: 2,
-        paddingVertical: 8,
-        paddingHorizontal: 8,
-        backgroundColor: colors.uranus500,
-        borderColor: colors.uranus500,
-      })
-    );
+    const button = getByTestId('IconButton.Icon');
+    expect(button.props.children.props.color).toEqual(colors.solidBrightLightest);
+    expect(button.props.children.props.size).toEqual(20);
```
Comment (lucianomlima) [inline src/components/Buttons/__tests__/IconButton.test.tsx:24]: Can use `toHaveProp` for both

File: src/components/Buttons/BaseButton.tsx [hunk 25]
```diff
@@ -40,24 +35,20 @@ const styles = StyleSheet.create({
...
-  loading = false,
-  disabled = false,
   activityIndicatorColor,
   children,
-  size = 'medium',
-  noHorizontalPadding = false,
   fill = false,
-  hasIcon = false,
+  isIconButton = false,
+  loading = false,
```
Comment (lucianomlima) [inline src/components/Buttons/BaseButton.tsx:42]: Optional props are false by default

File: src/components/Buttons/types.ts [hunk 26]
```diff
@@ -1,40 +1,46 @@
...
-import type { Size } from '@tokens/sizes';
+import type { Size } from '@components/types';
 
-export type ButtonSize = Size;
+export const buttonTypeOptions = ['solid', 'subtle', 'outline', 'ghost'] as const;
+export type ButtonType = typeof buttonTypeOptions[number];
 
-export type ButtonColor = 'uranus' | 'earth' | 'venus' | 'mars' | 'moon' | 'space';
+export const buttonVariantOptions = ['primary', 'secondary', 'alert', 'success', 'inversed', 'disabled'] as const;
+export type ButtonVariant = typeof buttonVariantOptions[number];
```
Comment (lucianomlima) [inline src/components/Buttons/types.ts:10]: Aí sim! Ficou top!

File: src/components/types.ts [hunk 27]
```diff
@@ -1,3 +1,6 @@
 import type { Insets } from 'react-native';
 
 export type HitSlop = null | Insets | number;
+
+export const sizeOptions = ['small', 'medium', 'large', 'xlarge'] as const;
+export type Size = typeof sizeOptions[number];
```
Comment (lucianomlima) [inline src/components/types.ts:6]: 👌🏼

**Review by lucianomlima — COMMENTED** [review]
Pronto, tinha deixado alguns comentários antes e finalizo com mais alguns agora.

File: storybook/options.ts (hunk 6 above)
Comment (lucianomlima) [reply storybook/options.ts]: 👏🏼

**Pull request merged**